Sources/SimpleCurrencyConverter/SimpleCurrencyConverter.swift

```diff
@@ -1,6 +1,138 @@
-public struct SimpleCurrencyConverter {
-    public private(set) var text = "Hello, World!"
+import Foundation
 
-    public init() {
+public enum Currency: String {
+    case JPY = "JPY"
+    case USD = "USD"
+    case GBP = "GBP"
+    case EUR = "EUR"
+}
+
+public class SimpleCurrencyConverter: CurrencyConverterProtocol {
+
+    private var exchangeRatesAPIKey: String = ""
+    private var networkManager: ExchangeRatesNetworkManager?
+    private static var sharedInstance: CurrencyConverterProtocol?
+    
+    /// Public setup singleton instance method for a class that allows for the setting of an exchange rates API key and a URL session.
+    ///
+    /// If an exchange rates API key is provided, it is set to the class's exchangeRatesAPIKey property.
+    /// If no exchange rates API key is provided, the class attempts to retrieve an API key from a plist file using the ExchangeRatesAPIKey keyword.
+    ///
+    /// If a URL session is provided, it is used to initialize the class's network manager.
+    /// If no URL session is provided, the class's network manager is initialized with the shared URL session.
+
+    public static func shared() -> CurrencyConverterProtocol {
+        if sharedInstance == nil {
+            fatalError("❗ Shared instance is used without framework initialization ❗")
+        }
+        return sharedInstance!
+    }
+    
+    public static func setup(exchangeRatesAPIKey: String? = nil, urlSession: URLSession? = nil) -> CurrencyConverterProtocol {
+        let container = DependencyContainer()
+        sharedInstance = SimpleCurrencyConverter(exchangeRatesAPIKey: exchangeRatesAPIKey, urlSession: urlSession)
+        container.register(type: CurrencyConverterProtocol.self, component: sharedInstance!)
```

Comment:
I'm not sure I understand the purpose of `DependencyContainer` here.
You create a local instance of `DependencyContainer`, register a new type and then resolve it and return its value. Isn't it just easier to simply return `sharedInstance!`. The `DependencyContainer` looks like it's supposed to store  instances and share them among all classes. Are you planning to use `DependencyContainer` in more places?

Reply:
No, I don't have the plan to use it in more places. In that case which solution would be better? Simple return sharedInstance! instead of Dependency Container implementation? Or maybe better to try to work around with current solution and make some changes to Dependency Container? Is it a good idea to use Dependency Container with Singleton object? I'm starting to doubt.

Reply:
with singletons where each class can init itself independently - probably not. 
It's just important to pass all dependencies in init of given class (even if it's a singleton).
`DependencyContainer` also helps to keep all init and setup logic in one place but your project is simple enough to look clean without it

Reply:
Ok, got it. It seems that my singleton instance and DependencyContainer are different approaches to getting one purpose and using them together has not made much sense according to my relatively simple and straightforward project structure. I think the most reasonable approach is to remove my DependencyContainer and work with the framework as a singleton. I'll leave this as a last resort. But I really want to understand how it could work under such circumstances because I had no experience using DependencyContainer with Singletons. So I provide another implementation. In the setup method, I registered through DependencyContainer my framework object instance with possible init params. Then the same registered object (looks like in general, it works like a singleton) will return in SimpleCurrencyConverter.shared method using DependencyContainer.resolve method. Is this approach acceptable? Thanks!

Reply:
It sounds like an acceptable solution. Let's discuss that topic on Slack/Zoom

Reply:
Great! Ready to discuss this at any time.